### ISSUE TITLE
Extended Raft algorithm with witness support

### ISSUE THREAD

**Issue by joshuazh-x** [etcd-io/raft#133]
The Raft algorithm requires an odd number of servers to maintain a quorum, meaning a minimum of three for a single point of failure. This isn't an issue for large systems but can be challenging for budget-limited customers needing fewer servers [^etcd-ask]. 

Efforts have been made in both scholarly circles [^pirogue][^flexiraft] and commercial sectors [^tikv][^platform9][^spectro] to resolve this issue. However, all existing research and implementations for small scale clusters (with two servers for example) either depend on another HA solution or necessitate a standalone server as a witness, adding to deployment complexity and potential performance bottlenecks.

I hereby propose extended Raft algorithm, a variant of Raft algorithm, which is designed for clusters with regular servers and a single witness, minimizing data traffic and access to witness while maintaining all key Raft properties. The witness in this algorithm is very suitable for implementation as a storage object with various options such as NFS, SMB, or cloud storage.

The extended Raft algorithm is backward compatible with Raft, meaning any cluster running with Raft can be seamlessly upgraded to support witness. 

The correctness of the algorithm has been conclusively proven through a formal proof in https://github.com/joshuazh-x/extended-raft-paper. Besides that, we also validate the formal specification of the algorithm using TLC model checker.

Look forward to your suggestions and feedback. 

@pav-kv @serathius @ahrtr @tbg @Elbehery @erikgrinaker @lemmy

[^etcd-ask]: https://github.com/etcd-io/etcd/issues/8934#issuecomment-398175955
[^pirogue]: Pâris, Jehan-François, and Darrell DE Long. "Pirogue, a lighter dynamic version of the Raft distributed consensus algorithm." 2015 IEEE 34th International Performance Computing and Communications Conference (IPCCC). IEEE, 2015.
[^flexiraft]: Yadav, Ritwik, and Anirban Rahut. "FlexiRaft: Flexible Quorums with Raft." The Conference on Innovative Data Systems Research (CIDR). 2023.
[^tikv]: https://github.com/tikv/tikv
[^platform9]: https://platform9.com/blog/transforming-the-edge-platform9-introduces-highly-available-2-node-kubernetes-cluster/
[^spectro]: https://www.spectrocloud.com/blog/two-node-edge-kubernetes-clusters-for-ha-and-cost-savings

**Comment by serathius** [comment]
+1 to witness support. My main concern would be creation of a test plan to ensure correctness. I think the TLC model checker will be crucial here.

**Comment by joshuazh-x** [comment]
Yes, indeed. We need tests to ensure implementation correctness besides algorithm itself. In our PoC work on etcd, we managed to make all existing tests (e2e, integration, robust) run on cluster with single witness. But additional tests would be needed to cover witness specific functionalities.

**Comment by Dogacel** [comment]
Hi @joshuazh-x, great work. What is the latest status of this project? Did you publish your paper? I see that @ZhouJianMS had a draft PR but it never got merged. I'm interested in distributed system research so wanted to check in what is the latest status here and wanted to see if there are any resources I can follow.

Thanks in advance!

cc: @serathius in case there is any effort parallel to this

**Comment by serathius** [comment]
We got stuck on merging https://github.com/etcd-io/raft/pull/226, review or help split PR into smaller re-viewable chunks would be appriciated.

**Comment by joshuazh-x** [comment]
Sorry for the delay in following up on this issue. I've been tied up with another assignment since March.

@Dogacel - we're still actively working on the project, though progress has been slower than expected due to competing priorities. The paper is still in draft form but includes most of the key details. I’ll aim to update it next month and plan to submit it once it's ready. Jian’s PR is nearly complete, though it’s still missing a few bug fixes. I’ll see if I can share a more polished version soon.

@serathius - thanks so much for your interest. I’ve been working on refactoring the TLA+ model to make it easier to extend (e.g., to support features like async write), but I’m starting to think that might not be necessary. Instead, I’ll prepare another PR with some smaller improvements around trace validation.

By the way, we’ll be presenting our latest work on 2-node Kubernetes (based on this project) at [KubeCon 2025 Japan](https://kccncjpn2025.sched.com/event/1x716/2-node-kubernetes-a-reliable-and-compatible-solution-xin-zhang-guang-hu-microsoft?iframe=no). If you’re interested, we’d love to hear your feedback!

**Comment by Matts966** [comment]
Thank you for really interesting presentation!
My question is if it is possible to use cloud services not only local storage but also cloud storage and cloud databases as
witness node or is there any requirement for consistency & performance?

**Comment by kei01234kei** [comment]
> the witness functionality can also be incorporated into cloud storage, such as Azure Storage or AWS S3.

This statement in this paper might be the answer you're looking for :)
https://github.com/joshuazh-x/extended-raft-paper/blob/main/main.pdf

**Comment by Matts966** [comment]
@kei01234kei
Thank you!

**Comment by joshuazh-x** [comment]
Yes, like @kei01234kei mentioned You can use cloud storages like Azure Blob or AWS S3. But I would say you can still have more options. The key to witness implementation is to have all its states persisted after each operation, and write operation is always atomic so that each node can get the latest witness state and update it correctly. In our current experiments, we choose optimistic concurrency control simply because we don’t want to use flock in NFS due to its undesired performance, especially in failure conditions. So if you have a reliable distributed lock service and it does not cost you additionally, why not use it. But that means it will be your specific implementation and relies on your own.

I am preparing designs to review with etcd folks. I believe their expertise can help us find a best way to achieve this kind of requirements.

**Comment by Matts966** [comment]
@joshuazh-x 
Very extensible! Thank you!
